spring-cloud-stream-integration-tests/src/test/java/org/springframework/cloud/stream/config/StreamListenerTestAnnotatedArguments.java

```diff
@@ -0,0 +1,133 @@
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.handler.annotation.Header;
+import org.springframework.messaging.handler.annotation.Headers;
+import org.springframework.messaging.handler.annotation.Payload;
+
+/**
+ * @author Marius Bogoevici
+ * @author Ilayaperumal Gopinathan
+ */
+@RunWith(Parameterized.class)
+public class StreamListenerTestAnnotatedArguments {
+
+	private Class<?> configClass;
+
+	public StreamListenerTestAnnotatedArguments(Class<?> configClass) {
+		this.configClass = configClass;
+	}
+
+	@Parameterized.Parameters
+	public static Collection InputConfigs() {
+		return Arrays.asList(new Class[] {TestPojoWithAnnotatedArguments1.class, TestPojoWithAnnotatedArguments2.class});
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void testAnnotatedArguments() throws Exception {
+		ConfigurableApplicationContext context = SpringApplication.run(this.configClass,
+				"--server.port=0");
+
+		TestPojoWithAnnotatedArguments testPojoWithAnnotatedArguments = context
+				.getBean(TestPojoWithAnnotatedArguments.class);
+		Sink sink = context.getBean(Sink.class);
+		String id = UUID.randomUUID().toString();
+		sink.input().send(MessageBuilder.withPayload("{\"bar\":\"barbar" + id + "\"}")
+				.setHeader("contentType", "application/json").setHeader("testHeader", "testValue").build());
+		assertThat(testPojoWithAnnotatedArguments.receivedArguments).hasSize(3);
+		assertThat(testPojoWithAnnotatedArguments.receivedArguments.get(0)).isInstanceOf(FooPojo.class);
+		assertThat(testPojoWithAnnotatedArguments.receivedArguments.get(0)).hasFieldOrPropertyWithValue("bar",
+				"barbar" + id);
+		assertThat(testPojoWithAnnotatedArguments.receivedArguments.get(1)).isInstanceOf(Map.class);
+		assertThat((Map<String, String>) testPojoWithAnnotatedArguments.receivedArguments.get(1))
+				.containsEntry(MessageHeaders.CONTENT_TYPE, "application/json");
+		assertThat((Map<String, String>) testPojoWithAnnotatedArguments.receivedArguments.get(1))
+				.containsEntry("testHeader", "testValue");
+		assertThat(testPojoWithAnnotatedArguments.receivedArguments.get(2)).isEqualTo("application/json");
+		context.close();
+	}
+
+	@EnableBinding(Processor.class)
+	@EnableAutoConfiguration
+	public static class TestPojoWithAnnotatedArguments1 extends TestPojoWithAnnotatedArguments {
+
+		@StreamListener(Processor.INPUT)
+		public void receive(@Payload FooPojo fooPojo,
+				@Headers Map<String, Object> headers,
+				@Header(MessageHeaders.CONTENT_TYPE) String contentType) {
+			this.receivedArguments.add(fooPojo);
+			this.receivedArguments.add(headers);
+			this.receivedArguments.add(contentType);
+		}
+	}
+
+	@EnableBinding(Processor.class)
+	@EnableAutoConfiguration
+	public static class TestPojoWithAnnotatedArguments2 extends TestPojoWithAnnotatedArguments {
+
+		@StreamListener
+		public void receive(@Input(Processor.INPUT) @Payload FooPojo fooPojo,
+				@Headers Map<String, Object> headers,
```

Comment:
@markpollack https://github.com/spring-cloud/spring-cloud-stream/pull/681#discussion_r84966975

This is why I don't think we should have `@Input`/`@Output` for the 'classic' use case at all and reserve it for bindable types. See my comment above. The only concession in the name of consistency that we can make is to allow:

```
@StreamListener("foo") @SendTo("bar")
public Flux<String> toUpper(Flux<String> input) {
}
```

so that there a 1:1 match in the 1-input 0..1-output case, since that's fairly intuitive, and we could stick to the current approach where `@Input` is used only on method arguments, `@Output` only on arguments and return types.

@ilayaperumalg  when it comes to header extraction, the use of `@Headers` is a valid use case, but I don't find that  the use of `@Headers` with another parameter annotated with `@Input` is.